server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSInquireSPOptionsCheckUtility.java

```diff
@@ -0,0 +1,1881 @@
+          put(CICSLexer.IRC, ErrorSeverity.WARNING);
+          put(CICSLexer.LIST, ErrorSeverity.WARNING);
+          put(CICSLexer.MONITOR, ErrorSeverity.WARNING);
+          put(CICSLexer.NEXT, ErrorSeverity.WARNING);
+          put(CICSLexer.RRMS, ErrorSeverity.WARNING);
+          put(CICSLexer.RUNNING, ErrorSeverity.WARNING);
+          put(CICSLexer.SECDISCOVERY, ErrorSeverity.WARNING);
+          put(CICSLexer.SPECIAL, ErrorSeverity.WARNING);
+          put(CICSLexer.STANDARD, ErrorSeverity.WARNING);
+          put(CICSLexer.START, ErrorSeverity.WARNING);
+          put(CICSLexer.STATISTICS, ErrorSeverity.WARNING);
+          put(CICSLexer.STORAGE, ErrorSeverity.WARNING);
+          put(CICSLexer.STORAGE64, ErrorSeverity.WARNING);
+          put(CICSLexer.SUSPENDED, ErrorSeverity.WARNING);
+          put(CICSLexer.SYSTEM, ErrorSeverity.WARNING);
+          put(CICSLexer.TAG, ErrorSeverity.WARNING);
+          put(CICSLexer.TCPIP, ErrorSeverity.WARNING);
+          put(CICSLexer.TEMPSTORAGE, ErrorSeverity.WARNING);
+          put(CICSLexer.TRACEDEST, ErrorSeverity.WARNING);
+          put(CICSLexer.TRACEFLAG, ErrorSeverity.WARNING);
+          put(CICSLexer.TRACETYPE, ErrorSeverity.WARNING);
+          put(CICSLexer.UOWDSNFAIL, ErrorSeverity.WARNING);
+          put(CICSLexer.UOWENQ, ErrorSeverity.WARNING);
+          put(CICSLexer.VTAM, ErrorSeverity.WARNING);
+          put(CICSLexer.WEB, ErrorSeverity.WARNING);
+          put(CICSLexer.WLMHEALTH, ErrorSeverity.WARNING);
+        }
+      };
+
+  public CICSInquireSPOptionsCheckUtility(
+      DialectProcessingContext context, List<SyntaxError> errors) {
+    super(context, errors, DUPLICATE_CHECK_OPTIONS);
+  }
+
+  /**
+   * Entrypoint to check CICS SP Inquire rule options
+   *
+   * @param ctx ParserRuleContext subclass containing options
+   * @param <E> A subclass of ParserRuleContext
+   */
+  public <E extends ParserRuleContext> void checkOptions(E ctx) {
+    switch (ctx.getRuleIndex()) {
+      case CICSParser.RULE_cics_inquire_association_list:
+        CICSParser.Cics_inquire_association_listContext listContext =
+            (CICSParser.Cics_inquire_association_listContext) ctx;
+        checkHasMandatoryOptions(listContext.LIST(), ctx, "LIST");
+        checkHasMandatoryOptions(listContext.LISTSIZE(), ctx, "LISTSIZE");
+        break;
+
+      case CICSParser.RULE_cics_inquire_atomservice:
+        CICSParser.Cics_inquire_atomserviceContext atomserviceContext =
+            (CICSParser.Cics_inquire_atomserviceContext) ctx;
+        checkBrowseMutuallyExclusive(atomserviceContext);
+        if (!atomserviceContext.START().isEmpty() || !atomserviceContext.END().isEmpty()) {
+          checkBrowsingInvalidOptions(atomserviceContext, CICSParser.ATOMSERVICE);
+          checkBrowsingHasNotParameter(atomserviceContext, CICSParser.ATOMSERVICE);
+        } else checkStatementHasParameter(atomserviceContext, CICSParser.ATOMSERVICE);
```

Comment:
Would it be possible to handle this check via a common code path? Maybe another map that contains the details needed to execute the check?
Only the `CICSParser.ATOMSERVICE` part seems variable - maybe another map ruleindex->token?

Reply:
The `checkBrowseMutuallyExclusive` almost implements the check.